web/web/files--common/javascript/OZONE/utils.ts

```diff
@@ -0,0 +1,481 @@
+          // %B for full month name
+          .replace(/%B/g, monthNames[date.getMonth()])
+          // %c for local date representation
+          .replace(/%c/g, date.toLocaleString())
+          // %d for zero-filled day of the month
+          .replace(/%d/g, `${dayOfMonth}`.padStart(2, "0"))
+          // %e for day of the month
+          .replace(/%e/g, `${dayOfMonth}`)
+          // %H for hour 00-23
+          .replace(/%H/g, `${hours}`.padStart(2, "0"))
+          // %I for hour 00-12
+          .replace(/%I/g, `${hours12}`.padStart(2, "0"))
+          .replace(/%m/g, `${month + 1}`.padStart(2, "0"))
+          // %M for minutes
+          .replace(/%M/g, `${minute}`.padStart(2, "0"))
+          .replace(/%p/g, (hours < 12) ? 'AM' : 'PM')
+
+          .replace(/%S/g, `${seconds}`.padStart(2, "0"))
+          .replace(/%y/g, `${year}`.substring(2))
+          .replace(/%Y/g, `${year}`);
+
+        if (dateString.match(/%z/i)) {
+          // try to get zone from locale date string
+          let zone = date.toLocaleString().replace(/^.*?([A-Z]{3,}(?:\+[0-9]+)?).*$/, '$1');
+          if (zone === date.toLocaleString()) {
+            // If that failed, then just get the timezone offset and display
+            // that instead
+            let zoneoffset = date.getTimezoneOffset();
+            zoneoffset = -zoneoffset / 60;
+            zone = `${zoneoffset > 0 ? "+" : "-"}${zoneoffset < 10 ? "0" : ""}${zoneoffset}`;
+          }
+          dateString = dateString.replace(/%z/ig, zone);
+        }
+        if (dateString.match(/%O/) || options.match(/agohover/)) {
+          // time ago
+          let secAgo = OZONE.request.timestamp - timestamp;
+          secAgo += Math.floor(
+            ((new Date()).getTime() - OZONE.request.date.getTime()) / 1000
+          );
+          const agoString = OZONE.utils.calculateDateAgo(secAgo);
+
+          dateString = dateString.replace(/%O/, agoString);
+          if (options.match(/agohover/)) {
+            // TODO Localisation
+            const hovertext = `${agoString} ago`;
+            OZONE.dialog.hovertip.makeTip(
+              dateElement, { text: hovertext, style: { width: 'auto' } }
+            );
+            // dateElement is now a HovertipElement
+            YAHOO.util.Event.addListener(dateElement, 'mouseover', function (
+              this: HovertipElement,
+              _event: Event
+            ) {
+              // Yahoo sets this to the scope element (dateElement)
+              if (this.hovertip) {
+                // XXX Shouldn't be necessary when hovertip isn't optional
+                this.hovertip.getElementsByTagName('div')[0].innerHTML = `${agoString} ${ogettext('ago')}`;
+              }
+            });
+          }
+        }
+      }
+      if (dateString) {
+        dateElement.innerHTML = dateString;
+        dateElement.style.visibility = 'visible';
+        dateElement.style.display = 'inline';
+      }
+    });
+  },
```

Comment:
A lot of this stuff should either be in some shared TS library or in the backend